In [1]:
import pandas as pd
import numpy as np
import copy
from scipy import stats
from sklearn.metrics.pairwise import cosine_similarity
import operator
import matplotlib.pyplot as plt
%matplotlib inline
import warnings
warnings.filterwarnings('ignore')
from sklearn.metrics import roc_curve, auc

/Users/rajeevkumar/anaconda/lib/python3.4/site-packages/pandas/computation/__init__.py:19: UserWarning: The installed version of numexpr 2.4.4 is not supported in pandas and will be not be used

  UserWarning)


In [2]:
# output in a table containing player name, its traded age, predicted war, true war
final_output_hitters = pd.DataFrame()
final_output_pitchers = pd.DataFrame()

In [3]:
hitter_data = pd.read_csv('war_daily_bat.csv', sep=",")
pitcher_data = pd.read_csv('war_daily_pitch.csv', sep=",")
lahman_master = pd.read_csv('../baseballdatabank-master/core/Master.csv')

In [4]:
hitter_data.columns

Index(['name_common', 'age', 'mlb_ID', 'player_ID', 'year_ID', 'team_ID',
       'stint_ID', 'lg_ID', 'PA', 'G', 'Inn', 'runs_bat', 'runs_br', 'runs_dp',
       'runs_field', 'runs_infield', 'runs_outfield', 'runs_catcher',
       'runs_good_plays', 'runs_defense', 'runs_position', 'runs_position_p',
       'runs_replacement', 'runs_above_rep', 'runs_above_avg',
       'runs_above_avg_off', 'runs_above_avg_def', 'WAA', 'WAA_off', 'WAA_def',
       'WAR', 'WAR_def', 'WAR_off', 'WAR_rep', 'salary', 'pitcher', 'teamRpG',
       'oppRpG', 'oppRpPA_rep', 'oppRpG_rep', 'pyth_exponent',
       'pyth_exponent_rep', 'waa_win_perc', 'waa_win_perc_off',
       'waa_win_perc_def', 'waa_win_perc_rep', 'OPS_plus', 'TOB_lg', 'TB_lg'],
      dtype='object')

In [5]:
pitcher_data.columns

Index(['name_common', 'age', 'mlb_ID', 'player_ID', 'year_ID', 'team_ID',
       'stint_ID', 'lg_ID', 'G', 'GS', 'IPouts', 'IPouts_start',
       'IPouts_relief', 'RA', 'xRA', 'xRA_sprp_adj', 'xRA_def_pitcher', 'PPF',
       'PPF_custom', 'xRA_final', 'BIP', 'BIP_perc', 'RS_def_total',
       'runs_above_avg', 'runs_above_avg_adj', 'runs_above_rep',
       'RpO_replacement', 'GR_leverage_index_avg', 'WAR', 'salary', 'teamRpG',
       'oppRpG', 'pyth_exponent', 'waa_win_perc', 'WAA', 'WAA_adj',
       'oppRpG_rep', 'pyth_exponent_rep', 'waa_win_perc_rep', 'WAR_rep',
       'ERA_plus', 'ER_lg'],
      dtype='object')

In [6]:
# ages range [15,50], some are [52 to 58]
hitter_unique_ages = hitter_data['age'].unique()

In [7]:
table_hitter = pd.DataFrame()
table_pitcher = pd.DataFrame()
features_selected = ['name_common', 'mlb_ID', 'age', 'player_ID', 'year_ID', 'team_ID', 'WAR']

In [8]:
table_hitter = copy.deepcopy(hitter_data[features_selected])
table_pitcher = copy.deepcopy(pitcher_data[features_selected])
table_hitter = table_hitter.sort(['age'], ascending=[1])
table_hitter = table_hitter.dropna()
print(len(table_hitter))
table_pitcher = table_pitcher.sort(['age'], ascending=[1])
table_pithcer = table_pitcher.dropna()
print(len(table_pitcher))

102158
44950


In [39]:
current_age = [32]
future_trade_age = [35]

In [10]:
# find all players having data while at age = 30
# hitters_with_traded_age = table_hitter[table_hitter['age'] == future_trade_age]
# hitters_with_traded_age = np.array(hitters_with_traded_age['name_common'])
# hitters_with_traded_age

array(['Paul Splittorff', 'Tim Leary', 'Steve Lake', ..., 'Ellis Burks',
       'Alvaro Espinoza', 'Billy Shindle'], dtype=object)

In [11]:
list_of_player_current_age = np.array(table_hitter[table_hitter['age'] == current_age]['mlb_ID'])
len(list_of_player_current_age)

4771

In [50]:
def calculate_traded_vec(traded_player, current_age):
    # player [hitter] whose future trade value is to be calculated
    traded_player_vec = []
    ages = sorted(np.array(traded_player['age']))
    for age in range(16, int(current_age)+1):
        age = 1.0*age
        val = np.array(traded_player[traded_player['age'] == age]['WAR'])
        if len(val) > 0:
            traded_player_vec.append(max(val))
        else:
            traded_player_vec.append(0.01)
    mean_val = np.mean(traded_player_vec)
    var_val = np.std(traded_player_vec)
    traded_player_vec.append(mean_val)
    traded_player_vec.append(var_val)
    return traded_player_vec
    

In [151]:
def calculated_weighted_WAR(top_similar_players, future_trade_age):
    # cosine sim weighted future WAR value
    predicted_WAR = 0.0
    total_sim = 0.0
    for similar_player in top_similar_players:
        sim_score = similar_player[1][0][0]
        if sim_score > 0.75:
            player = similar_player[0]
            player_data = table_hitter[table_hitter['name_common'] == player]
            player_WAR_at_future_year = np.array(player_data[player_data['age'] == future_trade_age]['WAR'])
            if len(player_WAR_at_future_year) > 0: 
                player_WAR_at_future_year = max(player_WAR_at_future_year)
                total_sim += sim_score
                predicted_WAR += sim_score*player_WAR_at_future_year
    #         print(player, player_WAR_at_future_year)
    if total_sim == 0:
        final_predicted_WAR = 0.01
    else:
        final_predicted_WAR = predicted_WAR/total_sim
    print('Future predicted war is in range [ %f , %f ] in age %d for player %s' %(final_predicted_WAR -0.5, final_predicted_WAR + 0.5, future_trade_age, traded_player_name))
    return final_predicted_WAR

    
    
    

In [52]:
def calculated_similar_player_vec(traded_player_vec,traded_player_name, current_age, future_trade_age):
    list_player_vec = []
    player_dic_sim = {}
    for player in hitters_with_traded_age:
        if player != traded_player_name:
            player_vec = []
            player_data = table_hitter[table_hitter['name_common'] == player]
            for age in range(16, int(current_age)+1):
                age = 1.0*age
                val = np.array(player_data[player_data['age'] == age]['WAR'])
                if len(val) > 0:
                    player_vec.append(max(val))
                else:
                    player_vec.append(0.01)
            mean_val = np.mean(player_vec)
            var_val = np.std(player_vec)
            player_vec.append(mean_val)
            player_vec.append(var_val)
            list_player_vec.append(player_vec)
            cos_sim = cosine_similarity(traded_player_vec, player_vec)
            player_dic_sim[player] = cos_sim
    top_similar_players = sorted(player_dic_sim.items(), key=operator.itemgetter(1), reverse=True)
    output = calculated_weighted_WAR(top_similar_players, future_trade_age)
    return output


In [152]:
count = 100
player_names = []
player_actual_WAR = []
player_predicted_WAR = []
for i in range(62, count):
#     mlb_ID = list_of_player_current_age[i]
    mlb_ID = traded_MLB_IDS[i]
#     mlb_ID = 77582.0
    current_age_v = current_age[i]
    print(current_age_v)
    future_trade_age_v = future_trade_age[i]
    print(future_trade_age_v)
    traded_player = table_hitter[table_hitter['mlb_ID'] == mlb_ID]
    traded_player_name = np.array(traded_player['name_common'])
#     print(traded_player_name)
    traded_player_name = np.unique(traded_player_name)[0]
#     print(traded_player_name)
    
    traded_player_vec = calculate_traded_vec(traded_player, current_age_v)
    predicted_WAR = calculated_similar_player_vec(traded_player_vec,traded_player_name, current_age_v, future_trade_age_v)
    actual_WAR = np.array(traded_player[traded_player['age'] == future_trade_age_v]['WAR'])
    if len(actual_WAR) > 0:
        actual_WAR = max(actual_WAR)
    else:
        actual_WAR = 0.01
    player_names.append(traded_player_name)
    player_actual_WAR.append(actual_WAR)
    player_predicted_WAR.append(predicted_WAR)
final_output_hitters = pd.DataFrame({'Hitters': player_names, 'Actual WAR': player_actual_WAR, 'Predicted WAR': player_predicted_WAR})


33.0
35.0
Future predicted war is in range [ -0.490000 , 0.510000 ] in age 35 for player Cub Stricker
33.0
35.0
Future predicted war is in range [ -0.490000 , 0.510000 ] in age 35 for player Cub Stricker
32.0
34.0
Future predicted war is in range [ 0.840105 , 1.840105 ] in age 34 for player Yank Robinson
27.0
29.0
Future predicted war is in range [ 0.043225 , 1.043225 ] in age 29 for player Adonis Terry
20.0
22.0
Future predicted war is in range [ 1.521410 , 2.521410 ] in age 22 for player Joe Kelley
20.0
22.0
Future predicted war is in range [ 1.521410 , 2.521410 ] in age 22 for player Joe Kelley
26.0
28.0
Future predicted war is in range [ 2.633326 , 3.633326 ] in age 28 for player George Van Haltren
26.0
28.0
Future predicted war is in range [ 2.633326 , 3.633326 ] in age 28 for player George Van Haltren
33.0
35.0
Future predicted war is in range [ 0.665345 , 1.665345 ] in age 35 for player Cliff Carroll
30.0
32.0
Future predicted war is in range [ 0.022714 , 1.022714 ] in age 32 fo

In [192]:
len(player_names)

38

In [193]:
len(player_actual_WAR)

38

In [194]:
len(player_predicted_WAR)

38

In [195]:
final_output_hitters = pd.DataFrame({'Hitters': player_names, 'Actual WAR': player_actual_WAR, 'Predicted WAR': player_predicted_WAR})

In [196]:
final_output_hitters

Actual WAR             Hitters  Predicted WAR
0         0.01        Cub Stricker       0.010000
1         0.01        Cub Stricker       0.010000
2         0.01       Yank Robinson       1.340105
3         0.45        Adonis Terry       0.543225
4         6.52          Joe Kelley       2.021410
5         6.52          Joe Kelley       2.021410
6         1.82  George Van Haltren       3.133326
7         1.82  George Van Haltren       3.133326
8         0.01       Cliff Carroll       1.165345
9         1.43           Joe Quinn       0.522714
10        0.01    Danny Richardson       2.570192
11        0.01         Kid Summers       0.785193
12        2.07          Buck Ewing       1.496432
13        4.46        George Davis       2.842299
14        0.01          Bill Eagan       1.502313
15        0.01        Jack McMahon       2.586005
16        3.60        Roger Connor       1.088229
17        0.01       John Stafford       0.848229
18       -2.05          Jack Boyle       1.025666
19        0.01       Jack Sharrott       0.859279
20        1.64        Duke Farrell       2.796940
21        0.35        Frank Killen       2.975519
22        0.01        Tim O'Rourke       2.477249
23        0.01        Tim O'Rourke       2.477249
24        0.01        Harry Taylor       2.415742
25        7.51     Hughie Jennings       2.493236
26        7.51     Hughie Jennings       2.493236
27        0.01        Tony Mullane       0.019216
28        0.01        Tony Mullane       0.019216
29        0.01          Piggy Ward       2.554725
30        0.01          Piggy Ward       2.554725
31       -0.18       Mike Sullivan       0.431587
32        0.89           Dummy Hoy       1.600812
33        0.01          King Kelly       1.129366
34        4.32        Mike Tiernan       2.256858
35        5.61        Mike Griffin       2.690676
36        0.77        Bill Merritt       3.229951
37       -0.14     George Treadway       1.883953

In [185]:
final_output_hitters = final_output_hitters.drop_duplicates()
final_output_hitters

Actual WAR             Hitters  Predicted WAR
0         0.01        Cub Stricker       0.010000
2         0.01       Yank Robinson       1.340105
3         0.45        Adonis Terry       0.543225
4         6.52          Joe Kelley       2.021410
6         1.82  George Van Haltren       3.133326
7         1.82  George Van Haltren       3.133326
8         0.01       Cliff Carroll       1.165345
9         1.43           Joe Quinn       0.522714
10        0.01    Danny Richardson       2.570192
11        0.01         Kid Summers       0.785193
12        2.07          Buck Ewing       1.496432
13        4.46        George Davis       2.842299
14        0.01          Bill Eagan       1.502313
15        0.01        Jack McMahon       2.586005
16        3.60        Roger Connor       1.088229
17        0.01       John Stafford       0.848229
18       -2.05          Jack Boyle       1.025666
19        0.01       Jack Sharrott       0.859279
20        1.64        Duke Farrell       2.796940
21        0.35        Frank Killen       2.975519
22        0.01        Tim O'Rourke       2.477249
23        0.01        Tim O'Rourke       2.477249
24        0.01        Harry Taylor       2.415742
25        7.51     Hughie Jennings       2.493236
27        0.01        Tony Mullane       0.019216
28        0.01        Tony Mullane       0.019216
29        0.01          Piggy Ward       2.554725
31       -0.18       Mike Sullivan       0.431587
32        0.89           Dummy Hoy       1.600812
33        0.01          King Kelly       1.129366
34        4.32        Mike Tiernan       2.256858
35        5.61        Mike Griffin       2.690676
36        0.77        Bill Merritt       3.229951
37       -0.14     George Treadway       1.883953

In [197]:
final_output_hitters['Current_Age'] = current_age[62:100]

In [191]:
len(final_output_hitters)

34

In [198]:
final_output_hitters['Future_Age'] = future_trade_age[62:100]

In [199]:
final_output_hitters['MLB_ID'] = traded_MLB_IDS[62:100]

In [200]:
final_output_hitters['Transaction ID'] =  transaction_IDS[62:100]

In [207]:
final_output_hitters = final_output_hitters.drop_duplicates(keep=False)
final_output_hitters = final_output_hitters.drop_duplicates(keep=False)
final_output_hitters

Actual WAR             Hitters  Predicted WAR  Current_Age  Future_Age  \
0         0.01        Cub Stricker       0.010000         33.0        35.0   
2         0.01       Yank Robinson       1.340105         32.0        34.0   
3         0.45        Adonis Terry       0.543225         27.0        29.0   
4         6.52          Joe Kelley       2.021410         20.0        22.0   
6         1.82  George Van Haltren       3.133326         26.0        28.0   
7         1.82  George Van Haltren       3.133326         26.0        28.0   
8         0.01       Cliff Carroll       1.165345         33.0        35.0   
9         1.43           Joe Quinn       0.522714         30.0        32.0   
10        0.01    Danny Richardson       2.570192         30.0        32.0   
11        0.01         Kid Summers       0.785193         25.0        27.0   
12        2.07          Buck Ewing       1.496432         33.0        35.0   
13        4.46        George Davis       2.842299         22.0        24.0   
14        0.01          Bill Eagan       1.502313         24.0        26.0   
15        0.01        Jack McMahon       2.586005         23.0        25.0   
16        3.60        Roger Connor       1.088229         35.0        37.0   
17        0.01       John Stafford       0.848229         23.0        25.0   
18       -2.05          Jack Boyle       1.025666         27.0        29.0   
19        0.01       Jack Sharrott       0.859279         23.0        25.0   
20        1.64        Duke Farrell       2.796940         26.0        28.0   
21        0.35        Frank Killen       2.975519         22.0        24.0   
22        0.01        Tim O'Rourke       2.477249         29.0        31.0   
23        0.01        Tim O'Rourke       2.477249         29.0        31.0   
24        0.01        Harry Taylor       2.415742         27.0        29.0   
25        7.51     Hughie Jennings       2.493236         24.0        26.0   
27        0.01        Tony Mullane       0.019216         34.0        36.0   
28        0.01        Tony Mullane       0.019216         34.0        36.0   
29        0.01          Piggy Ward       2.554725         26.0        28.0   
31       -0.18       Mike Sullivan       0.431587         26.0        28.0   
32        0.89           Dummy Hoy       1.600812         31.0        33.0   
33        0.01          King Kelly       1.129366         35.0        37.0   
34        4.32        Mike Tiernan       2.256858         26.0        28.0   
35        5.61        Mike Griffin       2.690676         28.0        30.0   
36        0.77        Bill Merritt       3.229951         23.0        25.0   
37       -0.14     George Treadway       1.883953         27.0        29.0   

      MLB_ID  Transaction ID  
0   122871.0         48231.0  
2   121325.0         48231.0  
3   123202.0         48231.0  
4   116926.0         48232.0  
6   123646.0         48232.0  
7   123646.0         48232.0  
8   112051.0         48233.0  
9   120845.0         48233.0  
10  121155.0         48234.0  
11  122979.0         48235.0  
12  113959.0         48235.0  
13  113106.0         48235.0  
14  113690.0         48235.0  
15  118816.0         48236.0  
16  112589.0         48236.0  
17  122634.0         48236.0  
18  111326.0         48236.0  
19  122068.0         48236.0  
20  114019.0         48237.0  
21  117056.0         48237.0  
22  120062.0         48238.0  
23  120062.0         48238.0  
24  123139.0         48238.0  
25  116528.0         48238.0  
27  119489.0         48239.0  
28  119489.0         48239.0  
29  123914.0         48239.0  
31  122961.0         48243.0  
32  116238.0         48243.0  
33  116953.0         48243.0  
34  123337.0         48243.0  
35  115143.0         48243.0  
36  118980.0         48244.0  
37  123450.0         48244.0

In [208]:
final_output_hitters.to_csv('Predicted_WAR_38_Hitters_1.csv')

### Results :-
##### Future predicted war is 2.107890 in age 32 for player Brady Clark and actual is 2.3
##### Future predicted war is 1.470445 in age 32 for player Mike Benjamin and actual is 2.07
Future predicted war is in range 1.34 in age 35 for player Ivan Rodriguez and actual is 1.36
Future predicted war is in range 1.7 in age 35 for player Mark McGwire and actual is 5.2

In [144]:
from sklearn.metrics import mean_squared_error
from math import sqrt

# new_final_output_hitters = final_output_hitters[final_output_hitters['Actual WAR'] != 0.01]
new_final_output_hitters = copy.deepcopy(final_output_hitters)
y_actual = np.array(new_final_output_hitters['Actual WAR'])
y_predicted = np.array(new_final_output_hitters['Predicted WAR'])
new_final_output_hitters['low_interval'] = new_final_output_hitters['Predicted WAR'] - 0.5
new_final_output_hitters['high_interval'] = new_final_output_hitters['Predicted WAR'] + 0.5
# y_predicted = np.array(new_final_output_hitters['low_interval'])
# y_predicted = np.array(new_final_output_hitters['high_interval'])
rms = sqrt(mean_squared_error(y_actual, y_predicted))
print('Root mean squared error is : ', rms)

Root mean squared error is :  1.6012861245929257


/Users/rajeevkumar/anaconda/lib/python3.4/site-packages/ipykernel/__main__.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  if __name__ == '__main__':


In [176]:
new_final_output_hitters[new_final_output_hitters['Actual WAR'] != 0.01]

Actual WAR            Hitters  Predicted WAR  low_interval  high_interval
2        -0.04        Mike Morgan      -0.085479     -0.585479       0.414521
3         0.04        Brett Tomko       0.338539     -0.161461       0.838539
5         0.29  Frank Catalanotto       1.345845      0.845845       1.845845
7         0.04      Ryan Franklin       1.385055      0.885055       1.885055
11       -0.31         Gene Baker       0.925335      0.425335       1.425335
17       -0.04       Willie Jones       1.528119      1.028119       2.028119
18        0.44     Harry Brecheen       0.208005     -0.291995       0.708005
20        0.21        Dave Parker       1.125239      0.625239       1.625239
29        8.97        Willie Mays       1.398361      0.898361       1.898361
31       -0.46          Paul Bako      -0.076725     -0.576725       0.423275
33       -0.02          Ted Wilks      -0.103710     -0.603710       0.396290
37        0.60          Carl Mays       1.607884      1.107884       2.107884
39        1.87    Blondie Purcell       1.337435      0.837435       1.837435
40       -0.31         Gene Baker       0.925335      0.425335       1.425335
46        0.21       John Marzano       0.052644     -0.447356       0.552644
47        3.60        Ozzie Smith       1.486355      0.986355       1.986355

In [182]:
table_hitter[table_hitter['name_common'] == 'Willie Mays']

name_common    mlb_ID   age player_ID  year_ID team_ID    WAR
58771  Willie Mays  118495.0  20.0  mayswi01     1951     NYG   3.94
58772  Willie Mays  118495.0  21.0  mayswi01     1952     NYG   1.25
58773  Willie Mays  118495.0  23.0  mayswi01     1954     NYG  10.61
58774  Willie Mays  118495.0  24.0  mayswi01     1955     NYG   9.03
58775  Willie Mays  118495.0  25.0  mayswi01     1956     NYG   7.58
58776  Willie Mays  118495.0  26.0  mayswi01     1957     NYG   8.29
58777  Willie Mays  118495.0  27.0  mayswi01     1958     SFG  10.24
58778  Willie Mays  118495.0  28.0  mayswi01     1959     SFG   7.79
58779  Willie Mays  118495.0  29.0  mayswi01     1960     SFG   9.48
58780  Willie Mays  118495.0  30.0  mayswi01     1961     SFG   8.72
58781  Willie Mays  118495.0  31.0  mayswi01     1962     SFG  10.50
58782  Willie Mays  118495.0  32.0  mayswi01     1963     SFG  10.60
58783  Willie Mays  118495.0  33.0  mayswi01     1964     SFG  11.02
58784  Willie Mays  118495.0  34.0  mayswi01     1965     SFG  11.22
58785  Willie Mays  118495.0  35.0  mayswi01     1966     SFG   8.97
58786  Willie Mays  118495.0  36.0  mayswi01     1967     SFG   4.34
58787  Willie Mays  118495.0  37.0  mayswi01     1968     SFG   6.24
58788  Willie Mays  118495.0  38.0  mayswi01     1969     SFG   3.05
58789  Willie Mays  118495.0  39.0  mayswi01     1970     SFG   5.20
58790  Willie Mays  118495.0  40.0  mayswi01     1971     SFG   6.34
58792  Willie Mays  118495.0  41.0  mayswi01     1972     SFG   0.20
58791  Willie Mays  118495.0  41.0  mayswi01     1972     NYM   1.63
58793  Willie Mays  118495.0  42.0  mayswi01     1973     NYM  -0.02

In [200]:
player = 'Matt Williams'
mlb_ID = [121358.0, 118743.0, 110183.0]

In [201]:
table_hitter[table_hitter['name_common'] == player]

name_common    mlb_ID   age  player_ID  year_ID team_ID   WAR
1314  Roberto Alomar  110183.0  20.0  alomaro01     1988     SDP  4.33
1315  Roberto Alomar  110183.0  21.0  alomaro01     1989     SDP  4.43
1316  Roberto Alomar  110183.0  22.0  alomaro01     1990     SDP  3.38
1317  Roberto Alomar  110183.0  23.0  alomaro01     1991     TOR  4.54
1318  Roberto Alomar  110183.0  24.0  alomaro01     1992     TOR  6.64
1319  Roberto Alomar  110183.0  25.0  alomaro01     1993     TOR  6.10
1320  Roberto Alomar  110183.0  26.0  alomaro01     1994     TOR  2.24
1321  Roberto Alomar  110183.0  27.0  alomaro01     1995     TOR  2.68
1322  Roberto Alomar  110183.0  28.0  alomaro01     1996     BAL  5.27
1323  Roberto Alomar  110183.0  29.0  alomaro01     1997     BAL  3.43
1324  Roberto Alomar  110183.0  30.0  alomaro01     1998     BAL  3.74
1325  Roberto Alomar  110183.0  31.0  alomaro01     1999     CLE  7.41
1326  Roberto Alomar  110183.0  32.0  alomaro01     2000     CLE  5.57
1327  Roberto Alomar  110183.0  33.0  alomaro01     2001     CLE  7.29
1328  Roberto Alomar  110183.0  34.0  alomaro01     2002     NYM  0.60
1330  Roberto Alomar  110183.0  35.0  alomaro01     2003     NYM -0.21
1329  Roberto Alomar  110183.0  35.0  alomaro01     2003     CHW  0.06
1331  Roberto Alomar  110183.0  36.0  alomaro01     2004     ARI  0.08
1332  Roberto Alomar  110183.0  36.0  alomaro01     2004     CHW -0.79

In [92]:
columns = ['playerID', 'retroID', 'bbrefID', 'nameGiven']
lahman_dataset = lahman_master[columns]
lahman_dataset.columns

Index(['playerID', 'retroID', 'bbrefID', 'nameGiven'], dtype='object')

In [93]:
trade_data = pd.read_csv('../tranDB/tran.txt', sep=",", header=None)
trade_data.columns = ['primary-date', 'time', 'approximate-indicator', 'secondary-date', 'approximate-indicator', 'transaction-ID',
'player', 'type', 'from-team', 'from-league', 'to-team', 'to-league', 'draft-type', 'draft-round', 'pick-number', 'info']
trade_data = trade_data[trade_data['type'] == 'T ']

In [94]:
def find_year(z):
    year = z[0:4]
    return int(year)
new_trade_data = pd.DataFrame()
new_trade_data['year_ID'] = trade_data['primary-date'].apply(lambda x : find_year(str(x)))
new_trade_data['transaction-ID'] = trade_data['transaction-ID']
new_trade_data['player'] = trade_data['player']

In [95]:
def map_playerID(x):
    if x in lahman_dic:
        return lahman_dic[x]
    else:
        return x
    
lahman_dic = lahman_dataset.set_index('retroID')['nameGiven'].to_dict()
lahman_name_dic = lahman_master.set_index('nameGiven')['bbrefID'].to_dict()
new_trade_data['player'] = new_trade_data['player'].apply(lambda x: map_playerID(x))
    

In [104]:
def map_playerID(x):
    if x in lahman_name_dic:
        return lahman_name_dic[x]
    else:
        return x
lahman_name_dic = lahman_master.set_index('bbrefID')['nameGiven'].to_dict()
table_hitter['player Name'] = table_hitter['player_ID'].apply(lambda x: map_playerID(x))

In [105]:
table_hitter['key'] = table_hitter['player Name']+'_'+table_hitter['year_ID'].apply(lambda x : str(x))

In [106]:
new_trade_data['key'] = new_trade_data['player'] +'_'+ new_trade_data['year_ID'].apply(lambda x: str(x))

In [107]:
possible_trades = {}
for row in new_trade_data.iterrows():
    possible_trades[row[1][3]] = row[1][1]


In [108]:
sample = table_hitter.head(5)

In [109]:
bool_trade = []
mlb_transactionID = []
for row in table_hitter.iterrows():
    if row[1][8] in possible_trades:
        bool_trade.append(1)
        mlb_transactionID.append(possible_trades[row[1][8]])
    else:
        bool_trade.append(0)
        mlb_transactionID.append(None)

In [110]:
table_hitter['Traded or not'] = bool_trade
table_hitter['MLB Transaction ID'] = mlb_transactionID

In [111]:
table_hitter

name_common    mlb_ID   age  player_ID  year_ID team_ID   WAR  \
67897       Joe Nuxhall  119872.0  15.0  nuxhajo01     1944     CIN  0.00   
52556      Leonidas Lee  117605.0  16.0    leele01     1877     STL  0.05   
22763    Jim Derrington  113286.0  16.0  derriji01     1956     CHW  0.04   
7600       Coonie Blank  111075.0  16.0  blankco01     1909     STL -0.04   
12902   Putsy Caballero  111846.0  16.0  cabalpu01     1944     PHI -0.20   
82230       Carl Scheib  121788.0  16.0  scheica01     1943     PHA -0.09   
10352         Jim Britt  111484.0  16.0  brittji01     1872     BRA -0.23   
71159      Frank Pearce  120328.0  16.0  pearcfr01     1876     LOU -0.04   
11144       Tommy Brown  111594.0  16.0  brownto02     1944     BRO -2.05   
60426       Roger McKee  118770.0  16.0  mckeero01     1943     PHI  0.03   
24430       Jacob Doyle  113529.0  16.0  doylejo01     1872     NAT -0.42   
60024     Willie McGill  118703.0  16.0  mcgilwi01     1890     CLE  0.24   
32361       Alex George  114701.0  16.0  georgal01     1955     KCA -0.28   
87870       Joe Stanley  122671.0  16.0  stanljo02     1897     WHS -0.03   
96584        Piggy Ward  123914.0  16.0   wardpi01     1883     PHI -0.11   
40825          Tom Hess  115868.0  16.0   hessto01     1892     BLN -0.03   
12903   Putsy Caballero  111846.0  17.0  cabalpu01     1945     PHI -0.13   
88725      Chuck Stobbs  122808.0  17.0  stobbch01     1947     BOS -0.02   
45052          Joey Jay  116507.0  17.0    jayjo01     1953     MLN -0.06   
63259       Ron Moeller  119219.0  17.0  moellro01     1956     BAL -0.02   
56035      Tony Madigan  118122.0  17.0  madigto01     1886     WHS -0.64   
92624      George Tomer  123386.0  17.0  tomerge01     1913     SLB -0.03   
17123         Ed Clough  112416.0  17.0  clouged01     1924     STL -0.32   
23923      Jim Donnelly  137200.0  17.0  donneji02     1884     KCC -0.11   
100728        Jim Woods  124562.0  17.0  woodsji01     1957     CHC  0.00   
19593   Willie Crawford  112808.0  17.0  crawfwi01     1964     LAD  0.16   
94416    Fred Van Dusen  123640.0  17.0  vandufr01     1955     PHI  0.03   
3179          Art Bader  110423.0  17.0  baderar01     1904     SLB -0.03   
78619     Ben Rochefort  121335.0  17.0  rochebe01     1914     PHA  0.03   
49184    Ed Kirkpatrick  117120.0  17.0  kirkped01     1962     LAA -0.15   
...                 ...       ...   ...        ...      ...     ...   ...   
67175       Phil Niekro  119786.0  47.0  niekrph01     1986     CLE  0.02   
65039       Jamie Moyer  119469.0  47.0  moyerja01     2010     PHI -0.08   
30072      Julio Franco  212297.0  47.0  francju01     2006     NYM -0.02   
27222      Johnny Evers  113949.0  47.0  eversjo01     1929     BSN  0.00   
1510       Nick Altrock  110201.0  47.0  altroni01     1924     WSH  0.10   
98796      Hoyt Wilhelm  124261.0  47.0  wilheho01     1970     CHC  0.00   
98797      Hoyt Wilhelm  124261.0  48.0  wilheho01     1971     ATL  0.00   
89112      Gabby Street  122865.0  48.0  streega01     1931     STL -0.03   
67177       Phil Niekro  119786.0  48.0  niekrph01     1987     CLE  0.01   
74892        Jack Quinn  120843.0  48.0  quinnja01     1932     BRO  0.04   
67176       Phil Niekro  119786.0  48.0  niekrph01     1987     ATL  0.00   
60245    Deacon McGuire  118738.0  48.0  mcguide01     1912     DET  0.04   
30074      Julio Franco  212297.0  48.0  francju01     2007     NYM -0.08   
30073      Julio Franco  212297.0  48.0  francju01     2007     ATL  0.00   
67178       Phil Niekro  119786.0  48.0  niekrph01     1987     TOR  0.00   
98798      Hoyt Wilhelm  124261.0  48.0  wilheho01     1971     LAD -0.04   
65040       Jamie Moyer  119469.0  49.0  moyerja01     2012     COL -0.01   
2905       Jimmy Austin  110389.0  49.0  austiji01     1929     SLB -0.03   
74893        Jack Quinn  120843.0  49.0  quinnja01     1933     CIN -0.02   
45266   Hughie Jennings  116528.0  49.0  jennihu01     1918     DET  0.00   
51927   

In [89]:
table_Traded_hitters = table_hitter[table_hitter['Traded or not'] == 1]

In [116]:
table_Traded_hitters = table_Traded_hitters.sort(['MLB Transaction ID'], ascending=[1])

In [117]:
table_Traded_hitters

name_common    mlb_ID   age  player_ID  year_ID team_ID   WAR  \
21996           Pat Deasley  113176.0  27.0  deaslpa01     1885     NYG  1.13   
18174           John Connor  112587.0  23.0  connojo02     1885     LOU -0.10   
18173           John Connor  112587.0  23.0  connojo02     1885     BUF -0.06   
75538           Toad Ramsey  120919.0  20.0  ramseto01     1885     LOU -0.26   
62667        Jocko Milligan  119136.0  23.0  millijo01     1885     PHA  2.19   
28581          John Fischer  114158.0  28.0    fishe01     1885     BUF -0.06   
68041           Tom O'Brien  119908.0  25.0  o'brito01     1885     BAL  0.10   
18311           Dick Conway  112612.0  20.0  conwadi01     1886     BAL  0.03   
18309           Bill Conway  112614.0  24.0  conwabi01     1886     BAL -0.03   
68105         Pat O'Connell  119920.0  25.0  o'conpa01     1886     BAL -0.99   
67075            Hugh Nicol  119777.0  28.0  nicolhu01     1886     STL -0.23   
9339             Jack Boyle  111326.0  20.0  boyleja01     1886     CIN -0.11   
60461         Alex McKinnon  118784.0  29.0  mckinal01     1886     SLM  3.41   
82648        Otto Schomberg  121855.0  21.0  schomot01     1886     PIT  1.74   
59388         Jim McCormick  118595.0  30.0  mccorji01     1887     PIT  0.21   
94444    George Van Haltren  123646.0  21.0  vanhage01     1887     CHC -0.84   
49806         Jimmy Knowles  117211.0  30.0  knowlji01     1887     NYP -0.26   
41424            Paul Hines  115966.0  32.0  hinespa01     1887     WHS  2.70   
31741           Gid Gardner  114606.0  28.0  gardngi01     1887     IND -0.13   
59927         Chippy McGarr  118685.0  24.0  mcgarch01     1887     PHA  1.62   
33285          Bill Gleason  114857.0  28.0  gleasbi01     1887     STL  0.12   
23914          Jim Donnelly  113448.0  21.0  donneji01     1887     WHS -0.32   
44255            John Irwin  116408.0  25.0  irwinjo01     1887     WHS  0.46   
62669        Jocko Milligan  119136.0  25.0  millijo01     1887     PHA  1.39   
56681             Fred Mann  118238.0  29.0   mannfr01     1887     CLE  1.32   
56682             Fred Mann  118238.0  29.0   mannfr01     1887     PHA  0.20   
97450            Curt Welch  124056.0  25.0  welchcu01     1887     STL  0.95   
87003          John Sowders  122553.0  20.0  sowdejo01     1887     IND -0.05   
31742           Gid Gardner  114606.0  29.0  gardngi01     1888     PHI  0.08   
31743           Gid Gardner  114606.0  29.0  gardngi01     1888     WHS  0.04   
...                     ...       ...   ...        ...      ...     ...   ...   
47468            Nate Karns  501992.0  27.0  karnsna01     2015     TBR  0.15   
62290           Brad Miller  543543.0  25.0  millebr02     2015     SEA  0.60   
27582        Danny Farquhar  543144.0  28.0  farquda01     2015     SEA  0.00   
64601        Logan Morrison  489149.0  27.0  morrilo01     2015     SEA -0.06   
101590          Chris Young  432934.0  36.0  youngch03     2015     KCR  0.12   
40685        Chris Herrmann  543302.0  27.0  herrmch01     2015     MIN  0.01   
40975           Aaron Hicks  543305.0  25.0  hicksaa01     2015     MIN  1.33   
65637      John Ryan Murphy  571974.0  24.0  murphjr01     2015     NYY  0.46   
67473            Aaron Nola  605400.0  22.0   nolaaa01     2015     PHI -0.10   
37791          Aaron Harang  421685.0  37.0  haranaa01     2015     PHI  0.15   
54104          Jose Lobaton  446653.0  30.0  lobatjo01     2015     WSN -0.55   
73324           Jose Pirela  517369.0  25.0  pireljo01     2015     NYY -0.57   
6232         Joaquin Benoit  276542.0  37.0  benoijo01     2015     SDP  0.04   
84950     Andrelton Simmons  592743.0  25.0  simmoan01     2015     ATL  4.01   
3061            Erick Aybar  430947.0  31.0  aybarer01     2015     LAA  2.31   
48812         Craig Kimbrel  518886.0  27.0  kimbrcr01     2015     SDP  0.02   
39617     Jeremy Hellickson  476451.0  28.0  hellije01     2015     ARI  0.31   
46587           James Jones  571825.0  26.0  jo

In [118]:
traded_MLB_IDS = np.array(table_Traded_hitters['mlb_ID'])

In [138]:
transaction_IDS = np.array(table_Traded_hitters['MLB Transaction ID'])

In [150]:
transaction_IDS[62:]

array([ 48231.,  48231.,  48231., ...,  77616.,  77618.,  77618.])

In [119]:
traded_MLB_IDS

array([ 113176.,  112587.,  112587., ...,  607359.,  521230.,  445926.])

In [120]:
current_age = np.array(table_Traded_hitters['age'])
future_trade_age = np.array(table_Traded_hitters['age'] + 2)

In [121]:
current_age

array([ 27.,  23.,  23., ...,  27.,  26.,  31.])

In [122]:
future_trade_age

array([ 29.,  25.,  25., ...,  29.,  28.,  33.])